controller/support_bundle_controller.go

```diff
@@ -0,0 +1,707 @@
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+const (
+	SupportBundleServiceAccount = "longhorn-support-bundle"
+
+	SupportBundleRequeueDelay = time.Second
+
+	SupportBundleMsgRequeueOnConflictFmt = "Requeue %v due to conflict"
+	SupportBundleMsgRequeueNextPhaseFmt  = "Requeue %v for next phase: %v"
+
+	SupportBundleMsgManagerPhase = "Support bundle manager updated phase: %v"
+
+	SupportBundleMsgDeleting = "Deleting support bundle manager"
+	SupportBundleMsgPurging  = "Purging failed support bundles"
+)
+
+type supportBundleRecordType string
+
+const (
+	supportBundleRecordNone   = supportBundleRecordType("")
+	supportBundleRecordError  = supportBundleRecordType("error")
+	supportBundleRecordNormal = supportBundleRecordType("normal")
+)
+
+type supportBundleRecord struct {
+	nextState longhorn.SupportBundleState
+
+	recordType supportBundleRecordType
+	reason     string
+	message    string
+}
+
+type SupportBundleController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID   string
+	serviceAccount string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+
+	httpClient rest.HTTPClient
+}
+
+func NewSupportBundleController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace, serviceAccount string) *SupportBundleController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	c := &SupportBundleController{
+		baseController: newBaseController("longhorn-support-bundle", logger),
+		controllerID:   controllerID,
+
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+
+		ds: ds,
+
+		kubeClient: kubeClient,
+		httpClient: &http.Client{Timeout: 30 * time.Second},
+
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: "longhorn-support-bundle-controller"}),
+	}
+
+	ds.SupportBundleInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Then, you need to change the code to `AddEventHandlerWithResyncPeriod` with 0 seconds of `resyncPeriod`, because, using `AddEventHandler` will just use the shared informer's resync period which is 30 seconds by default.

Reply:
Updated, thanks!